libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -835,67 +842,527 @@ bool GCS_MAVLINK::handle_mission_item(mavlink_message_t *msg, AP_Mission &missio
+void GCS_MAVLINK::remove_message_from_bucket(int8_t bucket, ap_message id)
 {
-    push_deferred_messages();
+    deferred_message_bucket[bucket].ap_message_ids.clear(id);
+    if (deferred_message_bucket[bucket].ap_message_ids.first_set() == -1) {
+        // bucket empty.  Free it:
+        deferred_message_bucket[bucket].interval_ms = 0;
+        deferred_message_bucket[bucket].last_sent_ms = 0;
+        if (sending_bucket_id == bucket) {
+            find_next_bucket_to_send();
+        }
+    }
 }
 
-// send a message using mavlink, handling message queueing
-void GCS_MAVLINK::send_message(enum ap_message id)
+bool GCS_MAVLINK::set_ap_message_interval(enum ap_message id, uint16_t interval_ms)
 {
-    uint8_t i, nextid;
+    if (id == MSG_NEXT_PARAM) {
+        // force parameters to *always* get streamed so a vehicle is
+        // recoverable from bad configuration:
+        if (interval_ms == 0) {
+            interval_ms = 100;
+        } else if (interval_ms > 1000) {
+            interval_ms = 1000;
+        }
+    }
 
-    if (id == MSG_HEARTBEAT) {
-        save_signing_timestamp(false);
+    // send messages out at most 80% of main loop rate
+    if (interval_ms != 0 &&
+        interval_ms*800 < AP::scheduler().get_loop_period_us()) {
+        interval_ms = AP::scheduler().get_loop_period_us()/800.0;
     }
 
-    // see if we can send the deferred messages, if any:
-    push_deferred_messages();
+    // check if it's a specially-handled message:
+    const int8_t deferred_offset = get_deferred_message_index(id);
+    if (deferred_offset != -1) {
+        deferred_message[deferred_offset].interval_ms = interval_ms;
+        deferred_message[deferred_offset].last_sent_ms = AP_HAL::millis16();
+        return true;
+    }
 
-    // if there are no deferred messages, attempt to send straight away:
-    if (num_deferred_messages == 0) {
-        if (try_send_message(id)) {
-            // yay, we sent it!
-            return;
+    // see which bucket has the closest interval:
+    int8_t closest_bucket = -1;
+    uint16_t closest_bucket_interval_delta = UINT16_MAX;
+    int8_t in_bucket = -1;
+    int8_t empty_bucket_id = -1;
+    for (uint8_t i=0; i<ARRAY_SIZE(deferred_message_bucket); i++) {
+        const deferred_message_bucket_t &bucket = deferred_message_bucket[i];
+        if (bucket.interval_ms == 0) {
+            // unused bucket
+            if (empty_bucket_id == -1) {
+                empty_bucket_id = i;
+            }
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+            if (bucket.ap_message_ids.first_set() != -1) {
+                ::fprintf(stderr, "Bucket %u has zero interval but with ids set", i);
+                abort();
+            }
+#endif
+            continue;
+        }
+        if (bucket.ap_message_ids.get(id)) {
+            in_bucket = i;
+        }
+        const uint16_t interval_delta = abs(bucket.interval_ms - interval_ms);
+        if (interval_delta < closest_bucket_interval_delta) {
```

Comment:
I think this should check that the bucket has a lower interval than requested: if user asks for 10ms intervals, I think they'll prefer 5ms intervals rather than 12.5ms. We should obviously count with the case of closest bucket not being set yet.

Reply:
That could lead to someone getting something at 400Hz when they asked for 10Hz (and all other messages bar one are at 4Hz)

Reply:
4Hz - 250ms interval
10Hz - 100ms interval
400Hz - 2.5ms interval

delta(4Hz - 10Hz) = 150ms
delta(400Hz - 10Hz) = 97.5ms

This means it already selects 400Hz with current code 😉 Even in an example that worked, in my opinion, that would be correct. If the user asks for a specific interval, it should be a maximum, higher intervals might not be useful at all (this would also mean decreasing the interval of a bucket if all are occupied with higher than requested intervals).